### PR TITLE
update benchmark ci

### LINKED ISSUE #5301: Add cadenced runs for benchmarking slicing
## Description

The goal of this task is to ensure that any existing benchmarking tools are running as expected. And that the output of those tasks can be easily used to detect performance changes in napari.

## Motivation

Slicing is a core part of napari and it's critical that any changes to the related code does not substantially negatively impact performance of slicing. Many of the existing ASV benchmarks measure the time and memory used during layer initialization and slicing, which could be affect by this project.

## Output

The key output of this task is to have some way of easily detecting when the performance of napari changes at some regular interval (e.g. daily, weekly) based on the existing ASV benchmarks. This could be as simple as using `asv publish` after running `asv continuous` in the existing GitHub workflow to generate static HTML that can be downloaded and rendered locally. Alternatively, the report could be entirely text based as long as it's easy to find any changes in performance.

We should also have some way to trigger a full benchmark run for a PR related to async slicing that also generates a similar report, but that could be done in a follow-up PR. This task can also be broken down as needed.

## Notes
need to add workflow runs for benchmarking

Jaime can support

Relevant previous PRs
- https://github.com/napari/napari/pull/4554
- https://github.com/napari/napari/pull/4656

<hr>

### KCP update:

Issues with the `Benchmark` workflow have been resolved in https://github.com/napari/napari/pull/5083

Unfortunately I introduced a regression which I'm fixing in https://github.com/napari/napari/pull/5246

### PR DESCRIPTION
**Pull request by kcpevey**
# Description

Benchmarks on CI had a few issues that were only uncovered once the updated benchmark_report.yml was committed to main (an unfortunate, known restriction - https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#workflow_run).

This PR aims to resolve the remaining issues.

## Type of change

- [x] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

Resolves #5301 

# How has this been tested?

I set up a complete duplication of the workflow on a new fork so that I could push changes directly to main and see the effects. (https://github.com/Quansight/napari-tmp/pull/1)

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/developers/translations.html).

### CODE REVIEW

File: .github/workflows/benchmarks.yml [hunk 1]
```diff
@@ -143,18 +143,21 @@ jobs:
...
           echo "${{ github.event.pull_request.number }}" > .asv/results/pr_number
-          cat <<EOF > .asv/results/message.txt
-          The benchmark run requested by $EVENT_NAME ($CONTENDER_REF vs $BASE_REF) has
-          finished with status `${{ steps.run_benchmark.outcome }}`. See the
+          cat <<EOF > .asv/results/message_${{ matrix.benchmark-name }}.txt
+          The ${{ matrix.benchmark-name }} benchmark run requested by $EVENT_NAME ($CONTENDER_REF vs $BASE_REF) has
+          finished with status '${{ steps.run_benchmark.outcome }}'. See the
           [CI logs and artifacts](||BENCHMARK_CI_LOGS_URL||) for further details.
+
           EOF
```
Comment (andy-sweet) [inline .github/workflows/benchmarks.yml]: Using `cat << EOF` creates newlines in the messages which looks weird on GitHub.

Screen Shot 2022-11-08 at 7 55 11 AM

Can we just use `echo` with newlines in the code but not in the output?

```suggestion
          echo \
          "The ${{ matrix.benchmark-name }} benchmark run requested by $EVENT_NAME ($CONTENDER_REF vs $BASE_REF) has" \
          "finished with status '${{ steps.run_benchmark.outcome }}'. See the" \
          "[CI logs and artifacts](||BENCHMARK_CI_LOGS_URL||) for further details." \
          > .asv/results/message_${{ matrix.benchmark-name }}.txt
```

**Review by andy-sweet — APPROVED** [review]
Looks good to me though that's mostly based on the successful run on https://github.com/Quansight/napari-tmp/pull/1

I had one minor comment on the message that should improve the formatting.

File: .github/workflows/benchmarks.yml (hunk 1 above)
Comment (kcpevey) [reply .github/workflows/benchmarks.yml]: Seems reasonable. I'll test and accept if it works as expected.

**Comment by kcpevey** [comment]
The Qt benchmarks are failing due to the old async benchmarks timing out. It also seems like the timeout which is set via the cli is not being honored or is being reported incorrectly. Unfortunately I can't reproduce this timeout on my own machine so I'm working in the dark a bit.

 We can provide a benchmark class-specific timeout. I'm attempting this approach while temporarily setting the benchmarks to "quick"

**Comment by kcpevey** [comment]
@andy-sweet the old async benchmarks are too slow to finish within the max github action runtime (6hrs). When the timeout for these was set at 300s, they ran in under 6 hrs but 12 of them hit the timeout. Some of these tests were taking 2-4 MINUTES. I bumped the timeout to 600s just to see if I could get them to finish, but this made us hit the github action timeout of 6 hours. 

How important is us to keep these in CI? I suggest that we skip these old async benchmarks in CI, but leave them in for now so we still use them for targeted comparisons (just not in CI while we're running the whole thing).

**Comment by andy-sweet** [comment]
> How important is us to keep these in CI? I suggest that we skip these old async benchmarks in CI, but leave them in for now so we still use them for targeted comparisons (just not in CI while we're running the whole thing).

What do you mean by old async benchmarks? Can you name a file or test suite name? If these are related to the experimental async code, then I agree we can skip them. If we introduced them as part of this new async effort, then I think we could either skip them temporarily or find ways to make them faster.

In general, I think we should make the benchmarks faster, while keeping them useful and representative of performance.

**Comment by kcpevey** [comment]
Oops, sorry I forgot the link: https://github.com/napari/napari/blob/main/napari/benchmarks/benchmark_qt_slicing.py

I don't actually know how to trigger the experimental async slicing, but the words "slicing" and "async" are used here so I made the assumption that was what this was testing...

**Comment by andy-sweet** [comment]
> Oops, sorry I forgot the link: https://github.com/napari/napari/blob/main/napari/benchmarks/benchmark_qt_slicing.py
> 
> I don't actually know how to trigger the experimental async slicing, but the words "slicing" and "async" are used here so I made the assumption that was what this was testing...

Yeah, we added these as part of the new effort in #4987. If there's a way to temporarily skip them from the workflow, then let's do that for that now. We can also raise a `NotImplementedError` in `setup` for those to skip them from the benchmarks themselves if needed.

**Pull request merged**